notifications_python_client/base.py

```diff
@@ -83,6 +83,16 @@ def _create_request_objects(self, url, data, params):
 
         return url, kwargs
 
+    def _serialize_data(self, data):
+        class ExtendedJSONEncoder(json.JSONEncoder):
```

Comment:
Is there a good reasons for re-defining the class each time the method is called?

Reply:
Hmm...the redefining didn't occur to me.

I wanted to keep the class close to the one place where it's used. It's also not meant to be part of the module namespace, so hiding it away in here seemed like a good solution.

Do you think there's a performance impact if we redefine it each time? One alternative might be to just move it directly above the method, so at least it's still nearby. How about that?

Reply:
I just did a toy example and found no performance difference.

<details><summary>Example A</summary>

```python

class Foo:

    def __init__(self):
        pass

    def foo(self):

        class Bar:

            def __init__(self):
                pass

            def bar():
                return 1 + 2

        return Bar().bar()

for i in range(1_000_000):
    Foo().foo()
```
</details>

<details><summary>Example B</summary>

```python

class Bar:

    def __init__(self):
        pass

    def bar():
        return 1 + 2

class Foo:

    def __init__(self):
        pass

    def foo(self):

        return Bar().bar()

for i in range(1_000_000):
    Foo().foo()
```
</details>

I think it’s more that when I see unpythonic<sup>1</sup> code like that I go ‘hmm, I wonder what edge case this developer was working around that meant they had to nest a class inside a method’. Whereas if you put the class in its own module I don’t have to wonder.

***

1. The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) says:
    > Simple is better than complex.
    >
    > Flat is better than nested.

Reply:
Thanks for checking the performance 🏅.

(I'd argue Python isn't as Pythonic as it used to be.)

It sounds like you'd find it less surprising to see this class in a separate module. Conversely, I'd find it more surprising: to me that implies it's "the" JSON encoder we use "for any JSON". I don't feel that strongly about it, but I was thinking a compromise of moving it just outside the method might be better all round.

Would you prefer it to be its own module, or would decreasing the nesting be OK?

Reply:
As far as I can see this is the only place we encode JSON in this repo therefore it is "the" JSON encoder for the Python client. I can’t see a future where we’d want the client using two different encoders.

Decreasing the nesting is OK though, since that makes it clear there’s no weird edge case that means the class needs to be redefined each time the method is called.

Reply:
OK I've amended the commit after also reading [the manual](https://docs.python.org/3/library/json.html#basic-usage) (😄) and realising there's a simpler way of doing it anyway, without having to define a class.

Reply:
> If specified, default should be a function that gets called for objects that can’t otherwise be serialized. It should return a JSON encodable version of the object or raise a TypeError. If not specified, TypeError is raised.

👍🏻